src/libstd/fs/mod.rs

```diff
@@ -571,7 +571,7 @@ pub fn create_dir<P: AsPath + ?Sized>(path: &P) -> io::Result<()> {
 pub fn create_dir_all<P: AsPath + ?Sized>(path: &P) -> io::Result<()> {
     let path = path.as_path();
     if path.is_dir() { return Ok(()) }
-    match path.parent() {
+    match path.without_file() {
```

Comment:
There's a special case below here testing for `file_name().is_none()`, and I think that the special case can be just entirely removed now in favor of `create_dir(path)`